appengine/pond_storage/rank.py

```diff
@@ -93,10 +99,36 @@ def create_ranking_match():
   # 4. Store match request in datastore.
   match = MatchRequest(entry_keys=entry_keys)
   match_key = match.put()
-  match_key_urlsafe = match_key.urlsafe()
-  logging.info('Created match request. matchKey:%s', match_key_urlsafe)
+  logging.info('Created match request. matchKey: %s', match_key)
   # 5. Return match request information.
-  return {'duck_list': duck_list, 'match_key': match_key_urlsafe}
+  return {'duck_list': duck_list, 'match_key': match_key.urlsafe()}
+
+def move_down_dummy_entry(dummy_entry, retries=5):
+  """Tries to move dummy entry down and/or deletes it if it is at bottom."""
+  @ndb.transactional(xg=True)
+  def swap_with_lower(lower_entry_key):
+    lower_entry = lower_entry_key.get()
+    # First confirm that lower entry rank is still farther down in leaderboard.
+    if not lower_entry and lower_entry.ranking > dummy_entry.key.get().ranking:
+      return False
+    return swap_order([lower_entry_key, dummy_entry.key])
+
+  leaderboard = dummy_entry.leaderboard_key.get()
+  if leaderboard.delete_if_last_entry(dummy_entry.key):
+    return
+  # Try to move down.
+  attempts = 0
+  while attempts < retries:
+    lower_entry_keys = (
+        leaderboard.get_entries_query().filter(
+            LeaderboardEntry.ranking > dummy_entry.key.get().ranking,
+            LeaderboardEntry.has_duck == True  # Ignore other dummy entries
+        )
+          .fetch(1, keys_only=True))
```

Comment:
Instead of having retry could we just fetch 5 ducks at once and then check to see if it is empty?